deepspeed/ops/transformer/inference/op_binding/mlp_gemm.py

```diff
@@ -40,35 +43,46 @@ def forward(self,
-                input,
-                residual,
-                input_bias,
-                weight_interm,
-                weight_out,
-                bias,
-                gamma,
-                beta,
-                self.config.epsilon,
-                self.config.pre_layer_norm,
-                self.config.mlp_after_attn,
-                weight_interm.scale if hasattr(weight_interm, 'scale') else torch.empty(1),  # type: ignore
-                weight_out.scale if hasattr(weight_out, 'scale') else torch.empty(1),  # type: ignore
-                self.config.dtype == torch.int8,
-                self.config.mlp_act_func_type,
-                self.config.transposed_mode)
+        if self.mlp_gemm_func != None:
+            if self.config.norm_type == NormType.LayerNorm:
+                output, residual_add = self.mlp_gemm_func(
+                    input,
+                    residual,
+                    input_bias,
+                    weight_interm,
+                    weight_out,
+                    bias,
+                    gamma,
+                    beta,
+                    self.config.epsilon,
+                    self.config.pre_layer_norm,
+                    self.config.mlp_after_attn,
+                    weight_interm.scale if hasattr(weight_interm, 'scale') else torch.empty(1),  # type: ignore
+                    weight_out.scale if hasattr(weight_out, 'scale') else torch.empty(1),  # type: ignore
+                    self.config.dtype == torch.int8,
+                    self.config.mlp_act_func_type,
+                    self.config.transposed_mode)
+            else:
+                output, residual_add = self.mlp_gemm_func(
+                    input,
+                    residual,
+                    weight_interm,
+                    weight_out,
+                    gamma,
+                    self.config.epsilon,
+                    weight_interm.scale if hasattr(weight_interm, 'scale') else torch.empty(1),  # type: ignore
+                    weight_out.scale if hasattr(weight_out, 'scale') else torch.empty(1),  # type: ignore
+                    self.config.dtype == torch.int8,
+                    self.config.mlp_act_func_type,
+                    self.config.transposed_mode)
         else:
-            output, residual_add = self.mlp_gemm_func(
-                input,
-                residual,
-                weight_interm,
-                weight_out,
-                gamma,
-                self.config.epsilon,
-                weight_interm.scale if hasattr(weight_interm, 'scale') else torch.empty(1),  # type: ignore
-                weight_out.scale if hasattr(weight_out, 'scale') else torch.empty(1),  # type: ignore
-                self.config.dtype == torch.int8,
-                self.config.mlp_act_func_type,
-                self.config.transposed_mode)
+            # fallback
+            if self.config.mlp_after_attn and not self.config.transposed_mode:
```

Comment:
We plan to guard fallback path with a environment variable and not triggered by default.  Fallback path does help for debugging when we enable kernels in new backend.   But it also potentially cause confusion, we observe people think inference kernels had been 'implemented' when fallback path is triggered, so we will guard these fallback path with environment variable.

Reply:
Fallback path are now guarded and will not be turned on by default.  Fallback path are used for debugging purpose, i.e. integrate one kernel before other kernels is available.